### PR TITLE
pymagicc.io refactor

### PR DESCRIPTION
**Pull request by lewisjared**
# Pull request

Please confirm that this pull request has done the following:

- [x] Description in ``CHANGELOG.rst`` added

Separated `pymagicc.io` from a module into a package. Each file type is now in its own module to make it easier to edit in future.

A future cleanup would be splitting the tests, but I tried to not touch them in the PR

### CODE REVIEW

**Comment by lewisjared** [comment]
@znicholls Any idea what's gone wrong with the window runners?

File: CHANGELOG.rst [hunk 1]
```diff
@@ -4,6 +4,7 @@ Changelog
 master
 ------
 
+- (`#303 <https://github.com/openscm/pymagicc/pull/303>`_) Refactor ``pymagicc.io`` into a package
```
Comment (znicholls) [inline CHANGELOG.rst:7]:
```suggestion
- (`#303 <https://github.com/openscm/pymagicc/pull/303>`_) Refactor ``pymagicc.io`` into multiple files
```

I'm not sure it's a package just cause it's in its own folder (not a big point but I think this use of language is more accurate)

File: pymagicc/io/__init__.py [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+from f90nml.namelist import Namelist
+from scmdata import ScmDataFrame
+
+from pymagicc.errors import NoReaderWriterError
+from pymagicc.magicc_time import convert_to_datetime
+from pymagicc.utils import _check_file_exists
+
+from .binout import _BinaryOutReader
+from .compact import _BinaryCompactOutReader, _CompactOutReader
+from .in_mon import (
```
Comment (znicholls) [inline pymagicc/io/__init__.py:18]:
```suggestion
from .in_files import (
```

what's `in_mon`

File: pymagicc/io/__init__.py [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+)
+from .mag import _MAGReader, _MAGWriter
+from .out import (
+    _EmisOutReader,
+    _InverseEmisReader,
+    _OutReader,
+    _TempOceanLayersOutReader,
+)
+from .prn import _PrnReader, _PrnWriter
+from .rcp import _RCPDatReader, _RCPDatWriter
```
Comment (znicholls) [inline pymagicc/io/__init__.py:38]:
```suggestion
from .rcpdat import _RCPDatReader, _RCPDatWriter
```

File: pymagicc/io/__init__.py [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+            "Couldn't find appropriate {} for {}.\nThe file must be one "
+            "of the following types and the filepath must match its "
+            "corresponding regular "
+            "expression:\n{}".format(tool_to_get, fbase, regexp_list_str)
+        )
+
+    raise NoReaderWriterError(error_msg)
+
+
+def _read_and_return_metadata_df(filepath):
```
Comment (znicholls) [inline pymagicc/io/__init__.py:252]:
```suggestion
def _read_metadata_and_df(filepath):
```

File: pymagicc/io/__init__.py [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+        "rcp85": "rcp85",
+    }
+    try:
+        return mapping[inname.lower()]
+    except KeyError:
+        error_msg = "No generic name for input: {}".format(inname)
+        raise ValueError(error_msg)
+
+
+def to_int(x):
```
Comment (znicholls) [inline pymagicc/io/__init__.py:442]: is this used anywhere? (probably out of this PR's scope but maybe worth making an issue to either a) move it into utils or b) delete)

File: pymagicc/io/__init__.py [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+            Passed to ``MAGICCData`` constructor (only used if ``other`` is not a
+            ``MAGICCData`` instance).
+
+        **kwargs
+            Passed to ``super().append()``
+        """
+        if not isinstance(other, MAGICCData):
+            other = MAGICCData(other, **constructor_kwargs)
+
+        if inplace:
```
Comment (znicholls) [inline pymagicc/io/__init__.py:584]: this metadata handling can be fixed by ScmRun at some point (again not for this PR but maybe move into a new issue?)

File: pymagicc/utils.py [hunk 7]
```diff
@@ -195,3 +196,8 @@ def get_date_time_string():
     """
     now = datetime.datetime.now()
     return now.strftime("%Y-%m-%d %H:%M")
+
+
+def _check_file_exists(file_to_read):
```
Comment (znicholls) [inline pymagicc/utils.py:201]: belongs in `io.utils` no?

**Comment by znicholls** [comment]
> A future cleanup would be splitting the tests, but I tried to not touch them in the PR

Nice yep let's leave for the next one

**Comment by znicholls** [comment]
> Any idea what's gone wrong with the window runners?

Nope, they won't even clone the repo...

File: CHANGELOG.rst (hunk 1 above)
Comment (lewisjared) [reply CHANGELOG.rst:7]: Python packages are any folder with an __init__.py, modules are files ending in `.py`

File: pymagicc/io/__init__.py (hunk 2 above)
Comment (lewisjared) [reply pymagicc/io/__init__.py:18]: This handles .IN/.MON files, but that naming works better

File: CHANGELOG.rst (hunk 1 above)
Comment (znicholls) [reply CHANGELOG.rst:7]: oh right oops I confused module and sub-package

File: pymagicc/io/__init__.py (hunk 5 above)
Comment (lewisjared) [reply pymagicc/io/__init__.py:442]: It's used in MAGICCData. Will likely disappear when that is updated to use ScmRun

**Comment by lewisjared** [comment]
@znicholls After some battling, I got it green

**Comment by znicholls** [comment]
One notebook ignore missing

**Comment by znicholls** [comment]
Tidy

**Pull request merged**